File: Assembler_Python/Question 3/symbol_table.py
import re

def create_symbol_table(asm_code):
    # Initialization
    symbol_table = {}
    pre_defined_symbols = {'SP': 0, 'LCL': 1, 'ARG': 2, 'THIS': 3, 'THAT': 4, 'SCREEN': 16384, 'KBD': 24576, 'R0': 0, 'R1': 1, 'R2': 2, 'R3': 3, 'R4': 4, 'R5': 5, 'R6': 6, 'R7': 7, 'R8': 8, 'R9': 9, 'R10': 10, 'R11': 11, 'R12': 12, 'R13': 13, 'R14': 14, 'R15': 15}
    symbol_table.update(pre_defined_symbols)
    address = 0

    # First pass
    lines = asm_code.split('\n')
    for line in lines :
        if '//' in line:
            line = line[:line.index('//')] #IGNORES THE COMMENTS
        if '(' in line:
            label = re.search(r'\(([^\s]+)\)', line).group(1)
            symbol_table[label] = address
        else:
            address += 1

    # Second pass
    address = 16
    for line in lines:
        if '//' in line:
            line = line[:line.index('//')]
        if '@' in line:
            symbol = re.search(r'@([^\s]+)', line).group(1)
            if symbol not in symbol_table:
                symbol_table[symbol] = address
                address += 1

    return symbol_table

asm_code = input("Please enter the asm file name including the extension: ")
with open(asm_code, 'r') as f:
    asm_code = f.read()

symbol_table = create_symbol_table(asm_code)
symbol_table_s = str(symbol_table)
output_file_name = 'symboltable.asm'
with open(output_file_name, 'w') as file:
    for symbol,value in symbol_table.items():
       file.write(f'{symbol}\t{value}\t\n')

File: Assembler_Java/Question_3/SymbolTable.java
import java.io.*;
import java.util.*;

public class SymbolTable {

    public static Map<String, Integer> createSymbolTable(String asmCode) {
        Map<String, Integer> symbolTable = new LinkedHashMap<>();

        int variableAddress = 16;

        // Predefined symbols
        symbolTable.put("SP", 0);
        symbolTable.put("LCL", 1);
        symbolTable.put("ARG", 2);
        symbolTable.put("THIS", 3);
        symbolTable.put("THAT", 4);
        symbolTable.put("SCREEN", 16384);
        symbolTable.put("KBD", 24576);
        for (int i = 0; i < 16; i++) {
            symbolTable.put("R" + i, i);
        }

        // First pass
        String[] lines = asmCode.split("\\n");
        int currentAddress = 0;
        for (String line : lines) {
            line = removeComments(line).trim();
            if (line.isEmpty()) {
                continue;
            }
            if (line.startsWith("(")) {
                String label = line.substring(1, line.length() - 1);
                symbolTable.put(label, currentAddress);
            } else {
                currentAddress++;
            }
        }

        // Second pass
        for (String line : lines) {
            line = removeComments(line).trim();
            if (line.isEmpty()) {
                continue;
            }
            if (line.startsWith("@")) {
                String symbol = line.substring(1);
                if (!symbolTable.containsKey(symbol)) {
                    symbolTable.put(symbol, variableAddress);
                    variableAddress++;
                }
            }
        }

        return symbolTable;
    }

    private static String removeComments(String line) {
        int index = line.indexOf("//");
        if (index != -1) {
            line = line.substring(0, index);
        }
        return line;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Please enter the asm file name including the extension: ");
        String asmCodeFileName = scanner.nextLine();
        scanner.close();

        try (BufferedReader fileReader = new BufferedReader(new FileReader(asmCodeFileName))) {
            StringBuilder asmCodeBuilder = new StringBuilder();
            String line;
            while ((line = fileReader.readLine()) != null) {
                asmCodeBuilder.append(line).append(System.lineSeparator());
            }
            String asmCode = asmCodeBuilder.toString();
            Map<String, Integer> symbolTable = createSymbolTable(asmCode);
            String outputFileName = "symboltable.asm";

            try (BufferedWriter fileWriter = new BufferedWriter(new FileWriter(outputFileName))) {
                for (Map.Entry<String, Integer> entry : symbolTable.entrySet()) {
                    fileWriter.write(entry.getKey() + "\t" + entry.getValue() + "\t\n");
                }
            } catch (IOException e) {
                System.err.println("Error writing to the output file: " + e.getMessage());
            }

            System.out.println("Symbol table has been created and saved to symboltable.asm.");
        } catch (IOException e) {
            System.err.println("An error occurred: " + e.getMessage());
        }
    }
}
